Closure: `variable require is undeclared` (regression since `2.0.23`)

### Stacktrace

```
 "/Users/aliced/Code/emsdk/node/14.15.5_64bit/bin/node" /Users/aliced/Code/emsdk/upstream/emscripten/tools/acorn-optimizer.js /var/folders/j9/lpk9tn816_v8fffzybq3wz140000gn/T/tmprpy83sxd.js applyDCEGraphRemovals --closureFriendly
 "/Users/aliced/Code/emsdk/node/14.15.5_64bit/bin/node" /Users/aliced/Code/emsdk/upstream/emscripten/tools/acorn-optimizer.js /var/folders/j9/lpk9tn816_v8fffzybq3wz140000gn/T/emscripten_temp_04wmh4_m/taisei.js.pp.js.jso.js.jso.js AJSDCE --closureFriendly
building:ERROR: Closure compiler run failed:

building:ERROR: tmph5lzeko1.js:195:4: ERROR - [JSC_UNDEFINED_VARIABLE] variable require is undeclared
  195|     require("fs").readFile(packageName, function(err, contents) {
           ^^^^^^^

1 error(s), 0 warning(s)

em++: error: closure compiler failed (rc: 1): /Users/aliced/Code/emsdk/node/14.15.5_64bit/bin/node --max_old_space_size=8192 /Users/aliced/Code/emsdk/upstream/emscripten/node_modules/.bin/google-closure-compiler --compilation_level ADVANCED_OPTIMIZATIONS --language_in ECMASCRIPT_2020 --language_out NO_TRANSPILE --emit_use_strict=false --externs tmpnk87kqug.js --js_output_file tmp7e7stvls.cc.js --js tmph5lzeko1.js the error message may be clearer with -g1 and EMCC_DEBUG=2 set
FAILED: src/taisei.html src/taisei.wasm src/taisei.js 
/Users/aliced/Code/emsdk/upstream/emscripten/em++ /Users/aliced/Code/dev-taisei/taisei.upstream/build/src/libtaisei-full.a --pre-js ../emscripten/preamble.js --pre-js resources/bundle_gfx.js --pre-js resources/bundle_misc.js --pre-js resources/bundle_bgm.js --pre-js resources/bundle_sfx.js --shell-file ../emscripten/shell.html -v -s STRICT=1 -Os -s ALLOW_MEMORY_GROWTH=1 -s 'DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=["$autoResumeAudioContext"]' -s ENVIRONMENT=web -s EXIT_RUNTIME=0 -s 'EXPORTED_FUNCTIONS=["_main", "_vfs_sync_callback"]' -s 'EXPORTED_RUNTIME_METHODS=["ccall"]' -s FILESYSTEM=1 -s FORCE_FILESYSTEM=1 -s GL_POOL_TEMP_BUFFERS=0 -s GL_PREINITIALIZED_CONTEXT=1 -s GL_SUPPORT_AUTOMATIC_ENABLE_EXTENSIONS=0 -s INITIAL_MEMORY=268435456 -s LZ4=1 -s MAX_WEBGL_VERSION=2 -s MIN_WEBGL_VERSION=2 -s MODULARIZE=0 -s WASM=1 -s LLD_REPORT_UNDEFINED -lGL -lwebgl.js -legl.js -lidbfs.js -lexports.js -s ASYNCIFY -Os -g0 -s ASSERTIONS=0 -s GL_TRACK_ERRORS=0 --closure 1 -o src/taisei.html
ninja: build stopped: subcommand failed.
```

### Relevant Links

See the following GitHub Action failure, running on Ubuntu: https://github.com/taisei-project/taisei/runs/2861465639

You can see the linker flags we set here (ignore the flags set under `if em_debug` as those are not used in our CI): https://github.com/taisei-project/taisei/blob/8fed29ca683b156cb3f6832a9bb2a9000b96cf32/src/meson.build#L163

Here is the `meson` machine file: https://github.com/taisei-project/taisei/blob/master/misc/ci/emscripten-build-test-ci.ini

Relevant section of our CI: https://github.com/taisei-project/taisei/blob/8fed29ca683b156cb3f6832a9bb2a9000b96cf32/.github/workflows/main.yml#L198

### Replicate

To replicate locally on the project itself, you will need `meson`.

Change `PATH_TO_EMSDK` below to reflect the correct path as an *absolute* path. Also, macOS users will need to use `gsed` (installed through Homebrew, etc). (Alternatively, set the `toolchain` variable inside `misc/ci/emscripten-build-test-ci.ini` to the *absolute* path to your `emsdk` installation, i.e: `toolchain = '/path/to/emsdk/upstream/emscripten/'` )

```bash
git clone https://github.com/taisei-project/taisei.git
cd taisei; git submodule update --init --recursive
sed -i "s+toolchain = ''+toolchain = 'PATH_TO_EMSDK/upstream/emscripten/'+g" misc/ci/emscripten-build-test-ci.ini # or gsed
meson setup build/ --cross-file=misc/ci/emscripten-build-test-ci.ini -Dbuild.cpp_std=gnu++14
meson compile -C build/
```

Tried with `tot` and it failed to compile as well. Reverting to `2.0.23` fixes this error.

If there is a new flag we have to add somewhere, please let us know.

----

As an aside, and I'll try to say this as politely as I can, Emscripten is easily the highest-burden platform we on the Taisei team support. It feels as if each minor version of Emscripten has made our builds fail in new and "interesting" ways, without much insight into why without spending hours digging into it, and despite our Emscripten support code not changing at all. We want to continue to support it, but it is so high-maintenance as to be a huge drag on our time and energy as an open source volunteer-run game project.

To illustrate this: we currently have *two* tests in our CI stack. One runs a "pinned" version of Emscripten (currently, `2.0.21`) and another runs with the `latest` Emscripten. It has become routine for minor versions of Emscripten to break our builds, and the `latest` test is set as `continue-on-error: true` as we expect it to fail often due to this. Perhaps we're using Emscripten in a completely unorthodox way, but I don't think that's the case, and it's usually the linking phase that has random issues like the one above.

While I recognize I am an outsider, and thus have no visibility into the inner machinations of this project, I highly recommend focusing on improving testing and communication around the closure compiler and any changes that might arrive in it, if that isn't already underway.

--- a/emcc.py
+++ b/emcc.py
@@ -2520,6 +2520,8 @@ def phase_source_transforms(options, target):
       file_args.append('--lz4')
     if options.use_preload_plugins:
       file_args.append('--use-preload-plugins')
+    if not settings.ENVIRONMENT_MAY_BE_NODE:
+      file_args.append('--no-node')
     file_code = shared.check_call([shared.FILE_PACKAGER, unsuffixed(target) + '.data'] + file_args, stdout=PIPE).stdout
     options.pre_js = js_manipulation.add_files_pre_js(options.pre_js, file_code)
 

--- a/tests/test_browser.py
+++ b/tests/test_browser.py
@@ -719,10 +719,17 @@ If manually bisecting:
     shutil.copyfile(test_file('screenshot.jpg'), 'screenshot.not')
     self.btest('sdl_image_prepare.c', reference='screenshot.jpg', args=['--preload-file', 'screenshot.not', '-lSDL', '-lGL'], also_proxied=True, manually_trigger_reftest=True)
 
-  def test_sdl_image_prepare_data(self):
+  @parameterized({
+    '': ([],),
+    # add testing for closure on preloaded files + ENVIRONMENT=web (we must not
+    # emit any node.js code here, see
+    # https://github.com/emscripten-core/emscripten/issues/14486
+    'closure_webonly': (['--closure', '1', '-s', 'ENVIRONMENT=web'],)
+  })
+  def test_sdl_image_prepare_data(self, args):
     # load an image file, get pixel data.
     shutil.copyfile(test_file('screenshot.jpg'), 'screenshot.not')
-    self.btest('sdl_image_prepare_data.c', reference='screenshot.jpg', args=['--preload-file', 'screenshot.not', '-lSDL', '-lGL'], manually_trigger_reftest=True)
+    self.btest('sdl_image_prepare_data.c', reference='screenshot.jpg', args=['--preload-file', 'screenshot.not', '-lSDL', '-lGL'] + args, manually_trigger_reftest=True)
 
   def test_sdl_image_must_prepare(self):
     # load an image file, get pixel data.

--- a/tools/file_packager.py
+++ b/tools/file_packager.py
@@ -21,7 +21,7 @@ data downloads.
 
 Usage:
 
-  file_packager TARGET [--preload A [B..]] [--embed C [D..]] [--exclude E [F..]]] [--js-output=OUTPUT.js] [--no-force] [--use-preload-cache] [--indexedDB-name=EM_PRELOAD_CACHE] [--separate-metadata] [--lz4] [--use-preload-plugins]
+  file_packager TARGET [--preload A [B..]] [--embed C [D..]] [--exclude E [F..]]] [--js-output=OUTPUT.js] [--no-force] [--use-preload-cache] [--indexedDB-name=EM_PRELOAD_CACHE] [--separate-metadata] [--lz4] [--use-preload-plugins] [--no-node]
 
   --preload  ,
   --embed    See emcc --help for more details on those options.
@@ -48,6 +48,8 @@ Usage:
 
   --use-preload-plugins Tells the file packager to run preload plugins on the files as they are loaded. This performs tasks like decoding images
                         and audio using the browser's codecs.
+
+  --no-node Whether to support Node.js. By default we do, which emits some extra code.
 
 Notes:
 
@@ -174,6 +176,7 @@ def main():
   separate_metadata = False
   lz4 = False
   use_preload_plugins = False
+  support_node = True
 
   for arg in sys.argv[2:]:
     if arg == '--preload':
@@ -203,6 +206,9 @@ def main():
       leading = ''
     elif arg == '--use-preload-plugins':
       use_preload_plugins = True
+      leading = ''
+    elif arg == '--no-node':
+      support_node = False
       leading = ''
     elif arg.startswith('--js-output'):
       jsoutput = arg.split('=', 1)[1] if '=' in arg else None
@@ -708,8 +714,10 @@ def main():
         }
       '''
 
-    ret += r'''
-      function fetchRemotePackage(packageName, packageSize, callback, errback) {
+    # add Node.js support code, if necessary
+    node_support_code = ''
+    if support_node:
+      node_support_code = r'''
         if (typeof process === 'object') {
           require('fs').readFile(packageName, function(err, contents) {
             if (err) {
@@ -720,6 +728,10 @@ def main():
           });
           return;
         }
+      '''
+    ret += r'''
+      function fetchRemotePackage(packageName, packageSize, callback, errback) {
+        %(node_support_code)s
         var xhr = new XMLHttpRequest();
         xhr.open('GET', packageName, true);
         xhr.responseType = 'arraybuffer';
@@ -770,7 +782,7 @@ def main():
       function handleError(error) {
         console.error('package error:', error);
       };
-    '''
+    ''' % { 'node_support_code' : node_support_code }
 
     code += r'''
       function processPackageData(arrayBuffer) {

--- a/tools/file_packager.py
+++ b/tools/file_packager.py
@@ -782,7 +782,7 @@ def main():
       function handleError(error) {
         console.error('package error:', error);
       };
-    ''' % { 'node_support_code' : node_support_code }
+    ''' % {'node_support_code': node_support_code}
 
     code += r'''
       function processPackageData(arrayBuffer) {

